### PR TITLE
[RDY] pygccxml + castxml + pybindgen

### PR DESCRIPTION
**Pull request by teto**
###### Motivation for this change
I use the simulator www.nsnam.org in my work. I need these libraries in order to generate the python bindings castxml + pygccxml. 

Castxml linking stage fails because it can't find -ltinfo. Comments in the repository say it's not available. Not sure what to do. 

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/compilers/ghc/7.4.2-binary.nix [hunk 1]
```diff
@@ -34,7 +34,7 @@ stdenv.mkDerivation rec {
     # GHC has dtrace probes, which causes ld to try to open /usr/lib/libdtrace.dylib
     # during linking
     stdenv.lib.optionalString stdenv.isDarwin ''
-      export NIX_LDFLAGS+=" -no_dtrace_dof"
+      export NIX_LDFLgGS+=" -no_dtrace_dof"
```
Comment (peti) [inline pkgs/development/compilers/ghc/7.4.2-binary.nix]: Huh? That's probably not intentional?

File: lib/licenses.nix [hunk 2]
```diff
@@ -39,6 +39,11 @@ lib.mapAttrs (n: v: v // { shortName = n; }) rec {
     url = http://developer.amd.com/amd-license-agreement/;
   };
 
+  apache2 = spdx {
+    spdxId = "AL-2.0";
```
Comment (peti) [inline lib/licenses.nix]: According to https://spdx.org/licenses/, that license does not exist.

File: pkgs/top-level/python-packages.nix [hunk 3]
```diff
@@ -1221,7 +1221,7 @@ in {
   };
 
   batinfo = buildPythonPackage rec {
-    version = "0.3";
+    version = "-1.3";
```
Comment (peti) [inline pkgs/top-level/python-packages.nix]: This is not right; the version must start with a digit.

**Review by peti — CHANGES_REQUESTED** [review]

File: lib/licenses.nix (hunk 2 above)
Comment (teto) [reply lib/licenses.nix]: Didn't know spdxId was an official register so I took a random one. Apache 2 license is registered though https://spdx.org/licenses/Apache-2.0.html. I will fix it

File: pkgs/development/compilers/ghc/7.4.2-binary.nix (hunk 1 above)
Comment (teto) [reply pkgs/development/compilers/ghc/7.4.2-binary.nix]: I switch between 3 different keyboards a day and sometimes I make mistakes. Will fix

File: pkgs/top-level/python-packages.nix (hunk 3 above)
Comment (teto) [reply pkgs/top-level/python-packages.nix]: typically typo mistake too. I will mark the PR as RFC once I make sure everything works. My last commit "Now works" was a shout of joy after sphalerite helped me solve sthg I 'd spent quite some time. thanks for the review

**Comment by teto** [comment]
tests fail because of an undesired warning generated by `-Wunused-command-line-argument`. Is that possible that the flag was added by cmake ? I looked at the cmake builder and could not find it. Might it be in clang-wrapped? tried in a nix-shell --pure with the same results. More details here
https://github.com/CastXML/CastXML/issues/90#issuecomment-337270818

File: pkgs/development/python-modules/pygccxml/default.nix [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  src = fetchFromGitHub {
+    owner  = "gccxml";
+    repo   = "pygccxml";
+    rev    = "v${version}";
+    sha256 = "02ip03s0vmp7czzflbvf7qnybibfrd0rzqbc5zfmq3zmpnck3hvm";
+  };
+
+  propagatedBuildInputs = [ pkgs.castxml ];
+
+  doCheck = false;
```
Comment (FRidh) [inline pkgs/development/python-modules/pygccxml/default.nix]: can you include a comment with a reason

File: lib/licenses.nix [hunk 5]
```diff
@@ -39,6 +39,11 @@ lib.mapAttrs (n: v: v // { shortName = n; }) rec {
     url = http://developer.amd.com/amd-license-agreement/;
   };
 
+  apache2 = spdx {
```
Comment (FRidh) [inline lib/licenses.nix]: This is the `asl20`

File: pkgs/development/tools/castxml/default.nix [hunk 6]
```diff
@@ -0,0 +1,70 @@
+{ pkgs
```
Comment (FRidh) [inline pkgs/development/tools/castxml/default.nix]: do not pass in `pkgs`

**Review by FRidh — CHANGES_REQUESTED** [review]

**Comment by teto** [comment]
I am trying to enable pygccxxml tests (previously disabled).

So first it generated
```
[nix-shell:~/pygccxml-v1.9.1-src]$ python -m unittests.test_all/home/teto/pygccxml-v1.9.1-src/unittests/data/core_types.hpp:9:10: fatal error: 
      'iostream' file not found
#include <iostream>
         ^~~~~~~~~~
1 error generated.
Traceback (most recent call last):
  File "/nix/store/2lb6nn8ivk1alhckv43n7734lqwbw7h9-python3-3.6.2/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/nix/store/2lb6nn8ivk1alhckv43n7734lqwbw7h9-python3-3.6.2/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/teto/pygccxml-v1.9.1-src/unittests/test_all.py", line 193, in <module>
    sys.exit(run_suite())
  File "/home/teto/pygccxml-v1.9.1-src/unittests/test_all.py", line 183, in run_suite
    result = unittest.TextTestRunner(verbosity=2).run(create_suite())
  File "/home/teto/pygccxml-v1.9.1-src/unittests/test_all.py", line 178, in create_suite
    main_suite.addTest(tester.create_suite())
  File "/home/teto/pygccxml-v1.9.1-src/unittests/decl_printer_tester.py", line 69, in create_suite
    suite.addTest(unittest.makeSuite(Test))
  File "/nix/store/2lb6nn8ivk1alhckv43n7734lqwbw7h9-python3-3.6.2/lib/python3.6/unittest/loader.py", line 503, in makeSuite
    testCaseClass)
  File "/nix/store/2lb6nn8ivk1alhckv43n7734lqwbw7h9-python3-3.6.2/lib/python3.6/unittest/loader.py", line 92, in loadTestsFromTestCase
    loaded_suite = self.suiteClass(map(testCaseClass, testCaseNames))
  File "/nix/store/2lb6nn8ivk1alhckv43n7734lqwbw7h9-python3-3.6.2/lib/python3.6/unittest/suite.py", line 24, in __init__
    self.addTests(tests)
  File "/nix/store/2lb6nn8ivk1alhckv43n7734lqwbw7h9-python3-3.6.2/lib/python3.6/unittest/suite.py", line 57, in addTests
    for test in tests:
  File "/home/teto/pygccxml-v1.9.1-src/unittests/decl_printer_tester.py", line 42, in __init__
    compilation_mode=parser.COMPILATION_MODE.FILE_BY_FILE)
  File "/home/teto/pygccxml-v1.9.1-src/pygccxml/parser/project_reader.py", line 264, in read_files
    return self.__parse_file_by_file(files)
  File "/home/teto/pygccxml-v1.9.1-src/pygccxml/parser/project_reader.py", line 292, in __parse_file_by_file
    decls = reader.read_file(header)
  File "/home/teto/pygccxml-v1.9.1-src/pygccxml/parser/source_reader.py", line 356, in read_file
    return self.read_cpp_source_file(source_file)
  File "/home/teto/pygccxml-v1.9.1-src/pygccxml/parser/source_reader.py", line 375, in read_cpp_source_file
    xml_file = self.create_xml_file(ffname)
  File "/home/teto/pygccxml-v1.9.1-src/pygccxml/parser/source_reader.py", line 324, in create_xml_file
    ": %s status:%s" % (gccxml_msg, exit_status))
RuntimeError: Error occurred while running CASTXML:  status:1
```

so I've tried to add differents packages such as` llvm.clangStdenv clang`: which found some headers but not all, stdarg.h seems to be clang specific yet and though I have add clang, it's gcc includes that seem to be used ? 
```
python -m unittests.test_allIn file included from /home/teto/pygccxml-v1.9.1-src/unittests/data/core_types.hpp:9:
In file included from /nix/store/rww78vdn2rkayrnqsjl8ib5iq2vfm3sw-gcc-6.4.0/include/c++/6.4.0/iostream:39:
In file included from /nix/store/rww78vdn2rkayrnqsjl8ib5iq2vfm3sw-gcc-6.4.0/include/c++/6.4.0/ostream:38:
In file included from /nix/store/rww78vdn2rkayrnqsjl8ib5iq2vfm3sw-gcc-6.4.0/include/c++/6.4.0/ios:38:
In file included from /nix/store/rww78vdn2rkayrnqsjl8ib5iq2vfm3sw-gcc-6.4.0/include/c++/6.4.0/iosfwd:40:
In file included from /nix/store/rww78vdn2rkayrnqsjl8ib5iq2vfm3sw-gcc-6.4.0/include/c++/6.4.0/bits/postypes.h:40:
In file included from /nix/store/rww78vdn2rkayrnqsjl8ib5iq2vfm3sw-gcc-6.4.0/include/c++/6.4.0/cwchar:44:
/nix/store/qxb0x9fcs60zv719yy0zm95gyd8mmgks-glibc-2.25-49-dev/include/wchar.h:40:11: fatal error: 
      'stdarg.h' file not found
# include <stdarg.h>
          ^~~~~~~~~~
1 error generated.
Traceback (most recent call last):
  File "/nix/store/2lb6nn8ivk1alhckv43n7734lqwbw7h9-python3-3.6.2/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/nix/store/2lb6nn8ivk1alhckv43n7734lqwbw7h9-python3-3.6.2/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/teto/pygccxml-v1.9.1-src/unittests/test_all.py", line 193, in <module>
    sys.exit(run_suite())
  File "/home/teto/pygccxml-v1.9.1-src/unittests/test_all.py", line 183, in run_suite

```
should I set CXXFLAGS myself or can a package do it for me ?

**Comment by Mic92** [comment]
Getting the included path for standard c++ headers is a bit tricky. The following expression might help:

```console
$ nix-instantiate --eval --expr 'with (import <nixpkgs>) {}; clang.default_cxx_stdlib_compile'
"-isystem $(echo -n /nix/store/x3wnjni3my5r8d8vpja3qsygvg4bcbgl-gcc-6.4.0/include/c++/*) -isystem $(echo -n /nix/store/x3wnjni3my5r8d8vpja3qsygvg4bcbgl-gcc-6.4.0/include/c++/*)/$(/nix/store/x3wnjni3my5r8d8vpja3qsygvg4bcbgl-gcc-6.4.0/bin/gcc -dumpmachine)"
```

**Comment by teto** [comment]
I changed the checkPhase to
```
   checkPhase = ''
    export CFLAGS="${clang.default_cxx_stdlib_compile}"
    echo $CFLAGS
    python -m unittests.test_all
  '';
```
but the code would still fail to find stdarg.h, I searched the echoed CFLAGS for stdarg.h
```
$ fd stdarg.h /nix/store/rww78vdn2rkayrnqsjl8ib5iq2vfm3sw-gcc-6.4.0/include/c++/6.4.0                                       
/nix/store/rww78vdn2rkayrnqsjl8ib5iq2vfm3sw-gcc-6.4.0/include/c++/6.4.0/tr1/stdarg.h
```
and it is there but in a subfolder. The homepage https://github.com/gccxml/pygccxml doesn't list clang compatibility (only python). Maybe I need to add a flag in CFLAG so that stdarg.h appears in the upper folder. Like -std=c++11 ?

**Comment by teto** [comment]
I tried enabling pygccxml tests but it proved kinda hard, environment variables are not useful so one needs to generate a file and the include_paths is checked so that `-isystem $(echo -n /nix/store/x3wnjni3my5r8d8vpja3qsygvg4bcbgl-gcc-6.4.0/include/c++/*) -isystem $(echo -n /nix/store/x3wnjni3my5r8d8vpja3qsygvg4bcbgl-gcc-6.4.0/include/c++/*)/$(/nix/store/x3wnjni3my5r8d8vpja3qsygvg4bcbgl-gcc-6.4.0/bin/gcc -dumpmachine)` is not ok.

Meanwhile it would be cool if it could be merged without the tests as I need it for my work.

PS: Here is what I had in my attempt to enable tests (for my future self or anyone interested in taking over)
```
  xml_generator = writeText "xml_generator.cfg" ''
    [xml_generator]
    # Specify which xml generator you want to use "castxml" or "gccxml"
    # "castxml is the default"
    xml_generator=castxml
    # Path to castxml or gccxml executable file
    xml_generator_path=${castxml}/bin/castxml
    # Set the path to the compiler (for example "/usr/bin/gcc") for CastXML
    compiler_path=${llvmPackages.clang-unwrapped}/bin/clang++
    # Gccxml working directory - optional, could be set to your source code directory
    working_directory=
    # Additional include directories, as list of paths ["path1/file1.h", "path2/file2.h", ...]
    include_paths=
    # You can explicitly set what compiler it should emulate (for GCCXML)
    # Valid options are: g++, msvc6, msvc7, msvc71, msvc8, cl.
    compiler=
    # Keep xml files after errors (useful for debugging)
    keep_xml=1

    '';

  checkPhase = ''
    # export CFLAGS="-I${llvmPackages.clang-unwrapped.out}/lib/clang/4.0.1/include"
 # - if [[ $CPPSTD == "-std=c++14" ]]; then
 #      mv unittests/configs/travis_std14.cfg unittests/xml_generator.cfg;
# fi
    # echo $CFLAGS
    cp ${xml_generator} unittests/xml_generator.cfg
    python3 -m unittests.test_all
  '';
debug  checkInputs = [which sphinx llvmPackages.libcxxClang pycodestyle llvmPackages.clang-unwrapped];
```

**Comment by teto** [comment]
@GrahamcOfBorg build pythonPackages.pybindgen pythonPackages.pygccxml castxml

**Comment by teto** [comment]
@GrahamcOfBorg build python3Packages.pybindgen python3Packages.pygccxml castxml

**Comment by teto** [comment]
I added pybindgen so as to support python bindings in https://github.com/NixOS/nixpkgs/pull/31346 (they released ns3.28 and fixed some problems in the build so I can finally update that PR). 
The maintainer of pygccxml clearly stated that the tests were tailored for its CI & when updating castxml, it went from 4 to 96 errors mostly because it checks the generated gcc command lines and nixos adds its own path -isystem ... that confuses the regexes. So I disabled the tests in both case.
I marked the PR as ready.

**Comment by Mic92** [comment]
Thanks!

**Pull request merged**